File: ts-sdk/src/alerts/run.handlers.on.alert.ts
import { Finding } from "../findings";
import { Logger, assertExists, assertFindings } from "../utils";
import { ScanAlertsOptions } from "../scanning";
import { Alert } from "./alert";
import { GetAlert } from "./get.alert";
import { CreateAlertEvent } from "./create.alert.event";
import { MetricsHelper } from "../metrics";

export type RunHandlersOnAlert = (
  alertOrHash: string | Alert,
  options: ScanAlertsOptions
) => Promise<Finding[]>;

export function provideRunHandlersOnAlert(
  getAlert: GetAlert,
  createAlertEvent: CreateAlertEvent,
  metricsHelper: MetricsHelper,
  shouldStopOnErrors: boolean,
  logger: Logger
): RunHandlersOnAlert {
  assertExists(getAlert, "getAlert");
  assertExists(createAlertEvent, "createAlertEvent");
  assertExists(metricsHelper, "metricsHelper");
  assertExists(logger, "logger");

  return async function runHandlersOnAlert(
    alertOrHash: string | Alert,
    options: ScanAlertsOptions
  ) {
    const { handleAlert } = options;
    if (!handleAlert) {
      throw new Error("no alert handler provided");
    }

    let alert: Alert;
    // if passed in a string hash
    if (typeof alertOrHash === "string") {
      logger.log(`fetching alert ${alertOrHash}...`);
      alert = await getAlert(alertOrHash);
    } else {
      // if passed in an alert
      alert = alertOrHash;
    }

    let findings: Finding[] = [];
    try {
      const alertEvent = createAlertEvent(alert);
      metricsHelper.startHandleAlertTimer(alert.hash!);
      findings = await handleAlert(alertEvent);
      metricsHelper.endHandleAlertTimer(alert.hash!);

      assertFindings(findings);
      logger.log(
        `${findings.length} findings for alert ${alert.hash} ${findings}`
      );
      metricsHelper.reportHandleAlertSuccess(findings.length);
    } catch (e) {
      metricsHelper.reportHandleAlertError();
      if (shouldStopOnErrors) {
        throw e;
      }
      logger.error(`${new Date().toISOString()}    handleAlert ${alert.hash}`);
      logger.error(e);
    }
    return findings;
  };
}

File: py-sdk/src/forta_bot_sdk/handlers/run_handlers_on_alert.py
from datetime import datetime
from typing import Callable, Optional
from ..alerts import Alert, CreateAlertEvent, GetAlert
from ..findings import Finding
from ..metrics import MetricsHelper
from ..common import ScanAlertsOptions
from ..utils import assert_exists, assert_findings, Logger, format_exception

RunHandlersOnAlert = Callable[[str | Alert,
                               ScanAlertsOptions, Optional[bool]], list[Finding]]


def provide_run_handlers_on_alert(
    get_alert: GetAlert,
    create_alert_event: CreateAlertEvent,
    metrics_helper: MetricsHelper,
    logger: Logger
) -> RunHandlersOnAlert:
    assert_exists(get_alert, 'get_alert')
    assert_exists(create_alert_event, 'create_alert_event')
    assert_exists(metrics_helper, 'metrics_helper')
    assert_exists(logger, 'logger')

    async def run_handlers_on_alert(
            alert_or_hash: str | Alert,
            options: ScanAlertsOptions,
            should_stop_on_errors: bool = True) -> list[Finding]:
        handle_alert = options.get('handle_alert')
        if not handle_alert:
            raise Exception("no alert handler provided")

        # if passed in a string hash
        if type(alert_or_hash) == str:
            logger.log(f'fetching alert {alert_or_hash}...')
            alert = await get_alert(alert_or_hash)
        else:
            # if passed in an alert
            alert = alert_or_hash

        findings = []
        try:
            alert_event = create_alert_event(alert)
            metrics_helper.start_handle_alert_timer(alert.hash)
            findings = await handle_alert(alert_event)
            metrics_helper.end_handle_alert_timer(alert.hash)

            assert_findings(findings)
            logger.log(
                f'{len(findings)} findings for alert {alert.hash} {findings if len(findings) > 0 else ""}')
            metrics_helper.report_handle_alert_success(len(findings))
        except Exception as e:
            metrics_helper.report_handle_alert_error()
            logger.error(
                f'{datetime.now().isoformat()}    handle_alert {alert.hash}')
            if should_stop_on_errors:
                raise e
            logger.error(format_exception(e))

        return findings

    return run_handlers_on_alert
